collector/hot_threads.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package collector
+
+import (
+	"fmt"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	defaultHotThreadsLabels = []string{"node", "thread_name", "thread_id"}
+
+	defaultHotThreadsLabelValues = func(HotThreads string) []string {
+		return []string{
+			HotThreads,
+		}
+	}
+	NODE_OUTPUT_SEPERATOR = ":::"
+	HOT_THREADS_OP_REGEX  = `^?([0-9]*[.])?[0-9]+%.*`
+	CPU_PERCENTAGE_REGEX  = `^?([0-9]*[.])?[0-9]+%`
+)
+
+// HotThreads information struct
+type HotThreads struct {
+	logger log.Logger
+	url    *url.URL
+
+	HotThreadsMetrics        HotThreadsMetric
+	HotThreadsFailureMetrics HotThreadsStepFailureMetric
+
+	jsonParseFailures prometheus.Counter
+}
+
+type HotThreadsMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(HotThreadsExp float64) float64
+	Labels func(HotThreadsDataNode string, HotThreadsName, HotThreadsId string) []string
+}
+
+type HotThreadsStepFailureMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(HotThreadsExp int64) float64
+	Labels func(HotThreadsIndex string, HotThreadsPolicy string, action string, step string) []string
+}
+
+func getEnv(key, defaultVal string) string {
```

Comment:
I don't think env vars should be read in a collector. All configuration and values should be passed in.